ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/Deduplicate.java

```diff
@@ -0,0 +1,141 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.decoder;
+
+import com.google.cloud.teleport.util.DurationUtils;
+import com.mozilla.telemetry.transforms.MapElementsWithErrors;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.UUID;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.options.ValueProvider.StaticValueProvider;
+import org.apache.beam.sdk.transforms.PTransform;
+import redis.clients.jedis.Jedis;
+
+/**
+ * Base class for deduplicating messages.
+ *
+ * <p>Subclasses are provided via static members removeDuplicates() and markAsSeen().
+ *
+ * <p>The packaging of subclasses here follows the style guidelines as captured in
+ * https://beam.apache.org/contribute/ptransform-style-guide/#packaging-a-family-of-transforms
+ */
+public abstract class Deduplicate extends MapElementsWithErrors.ToPubsubMessageFrom<PubsubMessage> {
+  private final ValueProvider<URI> uri;
+  private transient Jedis jedis;
+
+  Deduplicate(ValueProvider<URI> uri) {
+    this.uri = uri;
+  }
+
+  /**
+   * Lazy get transient {@link Jedis} client.
+   */
+  Jedis getJedis() {
+    if (jedis == null) {
+      jedis = new Jedis(uri.get());
+    }
+    return jedis;
+  }
+
+  /**
+   * Get {@code document_id} attribute from {@link PubsubMessage} as {@code byte[]}.
+   *
+   * @exception IllegalArgumentException if {@code document_id} is an invalid {@link UUID}.
+   */
+  static byte[] getId(PubsubMessage element) {
+    return Optional
+        .ofNullable(element.getAttributeMap())
+        .flatMap(m -> Optional.ofNullable(m.get("document_id")))
+        .map(UUID::fromString)
+        .map(id -> ByteBuffer
+            .wrap(new byte[16])
+            .putLong(id.getLeastSignificantBits())
+            .putLong(id.getMostSignificantBits())
+            .array())
+        .orElse(null);
```

Comment:
I think we could have this method return Optional<byte[]> and then not have to check for null later.